bitcoin/src/blockdata/script.rs

```diff
@@ -1311,22 +1342,87 @@ impl ScriptBuf {
+/// during serialization by `consensus_encode` and in the sighash cache we use `consensus_encode` to
+/// encode the script codes created below - therefore the length byte is not present in the raw
+/// script returned here. To get a hex string use `crate::consensus::serialize_hex(&script_code)`.
+impl ScriptBuf<ScriptCode> {
+    /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
+    /// for a P2WPKH output.
+    pub fn p2wpkh_script_code(wpkh: &Script) -> Option<Self> {
+        if !wpkh.is_v0_p2wpkh() {
+            return None
+        }
+        // From BIP143: For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac
+        let script = Builder::new()
+            // OP_PUSHBYTES_25 (0x19) is added by consensus_encode.
+            .push_opcode(OP_DUP)               // 0x76
+            .push_opcode(OP_HASH160)           // 0xa9
+            // OP_PUSHBYTES_20 (0x14) is added by push_slice
+            .push_slice(&wpkh.as_bytes()[2..]) // The `wpkh` script is 0x00, 0x14, <pubkey_hash>
+            .push_opcode(OP_EQUALVERIFY)       // 0x88
+            .push_opcode(OP_CHECKSIG)          // 0xac
+            .into_script();
+
+        Some(script)
+    }
+
+    /// Returns the script code used for spending a P2WSH output using `witness` (redeem script).
+    ///
+    /// # Errors
+    ///
+    /// If `witness` contains `OP_CODESEPARATOR` (currently unsupported).
+    pub fn p2wsh_script_code(witness: &Script) -> Result<Self, Error> {
+        // From BIP143:
+        // - if the witnessScript does not contain any OP_CODESEPARATOR, the scriptCode is the
+        // witnessScript serialized as scripts inside CTxOut.
+
+        // We explicitly do not support OP_CODESEPARATOR.
+        if witness.instructions().any(|x| {
+            // `matches!` isn't available in Rust 1.41.1
+            match x {
+                Ok(Instruction::Op(OP_CODESEPARATOR)) => true,
+                _ => false
+            }
+        }) {
+            // - if the witnessScript contains any OP_CODESEPARATOR, the scriptCode is the witnessScript
+            // but removing everything up to and including the last executed OP_CODESEPARATOR before the
+            // signature checking opcode being executed, serialized as scripts inside CTxOut. (The exact
+            // semantics is demonstrated in the examples below)
+            return Err(Error::FoundOpCodeseparator)
+        }
+
+        Ok(witness.to_owned())
+    }
+
+    /// Converts `raw_script` directly into a [`ScriptCode<ScriptCode>`].
+    ///
+    /// Assumes the given script is the correctly constructed script code. Useful for custom cases
+    /// with code separator, or in cases where the user wants to explicitly trust the given script as
+    /// a correctly computed script code.
+    pub fn dangerous_assume_script_code(raw_script: ScriptBuf) -> Self {
```

Comment:
In the light of recent discussion, this is not dangerous. But I think this will go away eventually and people will just convert from bytes.